File: ObjectDatabase/Local/src/main/java/org/elkoserver/objdb/ObjectStoreFactory.java
package org.elkoserver.objdb;

import org.elkoserver.foundation.boot.BootProperties;
import org.elkoserver.objdb.store.ObjectStore;
import org.elkoserver.util.trace.Trace;

import java.lang.reflect.InvocationTargetException;

public class ObjectStoreFactory {
    public static ObjectStore createAndInitializeObjectStore(BootProperties props, String propRoot, Trace trace) {
        ObjectStore result;

        String objectStoreClassName = props.getProperty(propRoot + ".objstore",
                "org.elkoserver.objdb.store.filestore.FileObjectStore");
        Class<?> objectStoreClass;
        try {
            objectStoreClass = Class.forName(objectStoreClassName);
        } catch (ClassNotFoundException e) {
            trace.fatalError("object store class " + objectStoreClassName + " not found");
            throw new IllegalStateException();
        }
        try {
            result = (ObjectStore) objectStoreClass.getConstructor().newInstance();
        } catch (IllegalAccessException e) {
            trace.fatalError("unable to access object store constructor: " + e);
            throw new IllegalStateException();
        } catch (InstantiationException e) {
            trace.fatalError("unable to instantiate object store object: " + e);
            throw new IllegalStateException();
        } catch (NoSuchMethodException e) {
            trace.fatalError("unable to find object store constructor: " + e);
            throw new IllegalStateException();
        } catch (InvocationTargetException e) {
            trace.fatalError("error during invocation of object store constructor: " + e.getCause());
            throw new IllegalStateException();
        }
        result.initialize(props, propRoot, trace);

        return result;
    }
}

File: ObjectDatabase/Local/src/main/kotlin/org/elkoserver/objdb/ObjectStoreFactory.kt
package org.elkoserver.objdb

import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.objdb.store.ObjectStore
import org.elkoserver.util.trace.slf4j.Gorgel
import java.lang.reflect.InvocationTargetException

object ObjectStoreFactory {
    fun createAndInitializeObjectStore(props: ElkoProperties, propRoot: String, baseGorgel: Gorgel): ObjectStore {
        val result: ObjectStore
        val objectStoreClassName = props.getProperty("$propRoot.objstore",
                "org.elkoserver.objdb.store.filestore.FileObjectStore")
        val objectStoreClass: Class<*>
        objectStoreClass = try {
            Class.forName(objectStoreClassName)
        } catch (e: ClassNotFoundException) {
            throw IllegalStateException("object store class $objectStoreClassName not found", e)
        }
        result = try {
            objectStoreClass.getConstructor().newInstance() as ObjectStore
        } catch (e: IllegalAccessException) {
            throw IllegalStateException("unable to access object store constructor", e)
        } catch (e: InstantiationException) {
            throw IllegalStateException("unable to instantiate object store object", e)
        } catch (e: NoSuchMethodException) {
            throw IllegalStateException("unable to find object store constructor", e)
        } catch (e: InvocationTargetException) {
            throw IllegalStateException("error during invocation of object store constructor", e)
        }
        result.initialize(props, propRoot, baseGorgel.getChild(objectStoreClass.kotlin))
        return result
    }
}